lib/src/rules/prefer_single_quotes.dart

```diff
@@ -0,0 +1,145 @@
+
+const _desc = "Prefer single quotes where they won't require escape sequences";
+
+const _details = '''
+
+**DO** use single quotes where they wouldn't require additional escapes.
+
+That means strings with an apostrophe may use double quotes so that the
+apostrophe isn't escaped (note: we don't lint the other way around, ie, a single
+quoted string with an escaped apostrophe is not flagged).
+
+Its also rare, but possible, to have strings within string interpolations. In
+this case, its much more readable to use a double quote somewhere. So double
+quotes are allowed either within, or containing, an interpolated string literal.
+Arguably strings within string interpolations should be its own type of lint.
+
+**BAD:**
+```
+useStrings(
+    "should be single quote",
+    r"should be single quote",
+    r"""should be single quotes""")
+```
+
+**GOOD:**
+```
+useStrings(
+    'should be single quote',
+    r'should be single quote",
+    r\'''should be single quotes\''',
+    "here's ok",
+    "nested \${a ? 'strings' : 'can'} be wrapped by a double quote",
+    'and nested \${a ? "strings" : "can be double quoted themselves"});
+```
+
+''';
+
+class PreferSingleQuotes extends LintRule {
+  _Visitor _visitor;
+  PreferSingleQuotes()
+      : super(
+            name: 'prefer_single_quotes',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitSimpleStringLiteral(SimpleStringLiteral string) {
+    if (string.isSingleQuoted || string.value.contains("'")) {
+      return;
+    }
+
+    // Bail out on 'strings ${x ? "containing" : "other"} strings'
+    if (!isNestedString(string)) {
+      rule.reportLintForToken(string.literal);
+    }
+  }
+
+  @override
+  visitStringInterpolation(StringInterpolation string) {
+    if (string.isSingleQuoted) {
+      return;
+    }
+
+    // slightly more complicated check there are no single quotes
+    if (string.elements
+        .any((e) => e is InterpolationString && e.value.contains("'"))) {
+      return;
+    }
+
+    // Bail out on "strings ${x ? 'containing' : 'other'} strings"
+    if (!containsString(string) && !isNestedString(string)) {
+      rule.reportLint(string);
+    }
+  }
+
+  /// Strings can be within interpolations (ie, nested). Check like this.
+  bool isNestedString(AstNode node) {
+    final checkWithinString = new _WithinStringVisitor();
+    node.parent?.accept(checkWithinString);
+
+    return checkWithinString.withinString;
+  }
+
+  /// Strings interpolations can contain other string nodes. Check like this.
+  bool containsString(StringInterpolation string) {
+    final checkHasString = new _HasStringVisitor();
+    for (final child in string.elements) {
+      child.accept(checkHasString);
```

Comment:
It probably won't matter much for performance, but you could short-circuit this loop by checking the value of `checkHasString.hasString` after each element rather than waiting to check until we've visited all of the elements.